### PR TITLE
[Evals API][6/n] meta-reference llm as judge, registration for ScoringFnDefs

### PR DESCRIPTION
**Pull request by yanxi0830**
--continuation of https://github.com/meta-llama/llama-stack/pull/323

# TL;DR
- Add LLM as judge meta-reference impl. Judge using Llama Stack inference_api. 
- Move ScoringFnDef to `*.json` based files for easier registration. 
- Support dynamic registering a ScoringFnDefs with LLMAsJudgeContext

#### Support Scoring Using a Judge
```python
response = await scoring_impl.score_batch(
    dataset_id=response[0].identifier,
    scoring_functions=[
        "meta-reference::llm_as_judge_8b_correctness",
    ],
)
```

#### Support Full Generate + Score using a Judge
```python
response = await eval_impl.evaluate_batch(
    dataset_id=response[0].identifier,
    candidate=ModelCandidate(
        model="Llama3.2-1B-Instruct",
        sampling_params=SamplingParams(),
    ),
    scoring_functions=[
        "meta-reference::subset_of",
        "meta-reference::llm_as_judge_8b_correctness",
    ],
  )
```

#### Support Register A Judge using ScoringFnDef
```python
# register the scoring function
await scoring_functions_impl.register_scoring_function(
  ScoringFnDefWithProvider(
      identifier="meta-reference::llm_as_judge_8b_random",
      description="Llm As Judge Scoring Function",
      parameters=[],
      return_type=NumberType(),
      context=LLMAsJudgeContext(
          prompt_template="""
          Output a number between 0 to 10. Your answer must match the format \n Number: <answer>
          """,
          judge_model="Llama3.1-8B-Instruct",
          judge_score_regex=[r"Number: (\d+)"],
      ),
      provider_id="test-meta",
  )
)

scoring_functions = await scoring_functions_impl.list_scoring_functions()
```

# Test

##### scoring using 3.1-8b judge
```
PROVIDER_ID=test-meta PROVIDER_CONFIG=llama_stack/providers/tests/scoring/provider_config_example.yaml pytest -s llama_stack/providers/tests/scoring/test_scoring.py --tb=short --disable-warnings
```
image

##### generate using 3.2-1b, score using 3.1-8b judge
```
PROVIDER_ID=test-meta PROVIDER_CONFIG=llama_stack/providers/tests/eval/provider_config_example.yaml pytest -s llama_stack/providers/tests/eval/test_eval.py --tb=short --disable-warnings
```
image

##### register a random judge function

image

```
results={
  'meta-reference::llm_as_judge_8b_random': ScoringResult(score_rows=[
    {
      'score': 5,
      'judge_feedback': 'Number: 5'
    },
    {
      'score': 5,
      'judge_feedback': 'Number: 5'
    },
    {
      'score': 5,
      'judge_feedback': 'Number: 5'
    },
    {
      'score': 5,
      'judge_feedback': 'Number: 5'
    },
    {
      'score': 7,
      'judge_feedback': 'Number: 7'
    }
  ],
  aggregated_results={
    'average': 5.4
  })
}
```

### CODE REVIEW

File: llama_stack/apis/scoring_functions/scoring_functions.py [hunk 1]
```diff
@@ -26,6 +26,10 @@ class Parameter(BaseModel):
 class LLMAsJudgeContext(BaseModel):
     judge_model: str
     prompt_template: Optional[str] = None
+    judge_score_regex: Optional[List[str]] = Field(
```
Comment (yanxi0830) [inline llama_stack/apis/scoring_functions/scoring_functions.py:29]: We are making a strong assumption that a prompt_template is associated with a judge_score_regex to extract judge scores. (referring to [braintrust templates](https://github.com/braintrustdata/autoevals/blob/613f300483ca05d1f7a5b59e07e4e4a1edcfe107/templates/factuality.yaml#L19))

File: llama_stack/providers/impls/meta_reference/scoring/scoring_fn/fn_defs/llm_as_judge_8b_correctness.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+Feedback:::
+Total rating: (your rating, as a int between 0 and 5)
+Now here are the question, expected_answer, system_answer.
+Question: {input_query}
+Expected Answer: {expected_answer}
+System Answer: {generated_answer}
+Feedback:::
+Total rating:
+"""
+llm_as_judge_8b_correctness_fn_def = ScoringFnDef(
```
Comment (ashwinb) [inline llama_stack/providers/impls/meta_reference/scoring/scoring_fn/fn_defs/llm_as_judge_8b_correctness.py:25]: nit: newline before this

another nit: maybe remove the `_fn_def` suffix?

File: llama_stack/providers/impls/meta_reference/scoring/scoring.py [hunk 3]
```diff
@@ -37,22 +38,39 @@ def __init__(
...
         self.datasets_api = datasets_api
-
-    async def initialize(self) -> None: ...
+        self.inference_api = inference_api
+        self.scoring_fn_id_impls = {}
+
+    async def initialize(self) -> None:
+        for x in FIXED_FNS:
+            impl = x()
+            await impl.initialize()
```
Comment (ashwinb) [inline llama_stack/providers/impls/meta_reference/scoring/scoring.py:52]: remove the `initialize()`

File: llama_stack/providers/impls/meta_reference/scoring/scoring.py [hunk 4]
```diff
@@ -37,22 +38,39 @@ def __init__(
...
+            for fn_defs in impl.get_supported_scoring_fn_defs():
+                self.scoring_fn_id_impls[fn_defs.identifier] = impl
+                self.llm_as_judge_fn = impl
 
     async def shutdown(self) -> None: ...
 
     async def list_scoring_functions(self) -> List[ScoringFnDef]:
-        return [x.scoring_function_def for x in SUPPORTED_SCORING_FNS]
+        return [
+            fn_defs
```
Comment (ashwinb) [inline llama_stack/providers/impls/meta_reference/scoring/scoring.py:66]: should be singular `fn_def`

**Review by ashwinb — APPROVED** [review]
looks good!

File: llama_stack/providers/impls/meta_reference/scoring/scoring_fn/fn_defs/llm_as_judge_8b_correctness.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    identifier="meta-reference::llm_as_judge_8b_correctness",
```
[the remark this answers is not in the document]
Comment (yanxi0830) [reply llama_stack/providers/impls/meta_reference/scoring/scoring_fn/fn_defs/llm_as_judge_8b_correctness.py:27]: Agreed, I have the braintrust scoring_fns following a very similar structure here: https://github.com/meta-llama/llama-stack/pull/333

**Pull request merged**